control-base/src/main/java/fi/nls/oskari/control/feature/SaveFeatureHandler.java

```diff
@@ -4,80 +4,70 @@
-            WFSLayerConfiguration lc = getWFSConfiguration(layer.getId());
+                OskariLayer layer = getLayer(featureJSON.optString("layerId"));
 
-            // TODO: rewrite to use wfs-t related code under myplaces OR atleast using an xml lib
-            StringBuilder requestData = new StringBuilder(
-                    "<wfs:Transaction service='WFS' version='1.1.0' " +
-                            "xmlns:ogc='http://www.opengis.net/ogc' " +
-                            "xmlns:wfs='http://www.opengis.net/wfs'>" +
-                            "<wfs:Update typeName='" + layer.getName() + "'>");
-            JSONArray jsonArray = jsonObject.getJSONArray("featureFields");
-            for (int i = 0; i < jsonArray.length(); i++) {
-                requestData.append("<wfs:Property><wfs:Name>" + jsonArray.getJSONObject(i).getString("key") +
-                        "</wfs:Name><wfs:Value>" + jsonArray.getJSONObject(i).getString("value") +
-                        "</wfs:Value></wfs:Property>");
-            }
+                if (!canEdit(layer, params.getUser())) {
+                    throw new ActionDeniedException("User doesn't have edit permission for layer: " + layer.getId());
+                }
+                ByteArrayOutputStream baos = new ByteArrayOutputStream();
 
-            if (jsonObject.has("geometries")) {
-                insertGeometries(lc.getGMLGeometryProperty(), requestData, jsonObject.getJSONObject("geometries"), srsName);
-            }
-            requestData.append("<ogc:Filter><ogc:FeatureId fid='" + jsonObject.getString("featureId") + "'/></ogc:Filter></wfs:Update></wfs:Transaction>");
+                List<Feature> features = new ArrayList<>();
+                Feature feature = getFeature(featureJSON);
+                features.add(feature);
+                FeatureWFSTRequestBuilder.updateFeatures(baos, features);
+
+                final String wfstMessage = baos.toString();
+                LOG.debug("Updating feature to service at", layer.getUrl(), "with payload", wfstMessage);
 
-            String responseString = postPayload(layer, requestData.toString());
-            flushLayerTilesCache(layer.getId());
+                String responseString = postPayload(layer, wfstMessage);
+                flushLayerTilesCache(layer.getId());
 
-            if (responseString.indexOf("Exception") > -1) {
-                ResponseHelper.writeResponse(params, "Exception");
-            } else if (responseString.indexOf("<wfs:totalUpdated>1</wfs:totalUpdated>") > -1) {
-                ResponseHelper.writeResponse(params, "");
+                if (responseString.indexOf("Exception") > -1) {
+                    throw new ActionException("Cannot save feature: " + responseString);
+                } else if (responseString.indexOf("<wfs:totalUpdated>1</wfs:totalUpdated>") > -1) {
+                    ResponseHelper.writeResponse(params, "Feature updated");
```

Comment:
We either throw an exception or write a response for each iteration of the loop.